traits/observers/_metadata_filter.py

```diff
@@ -0,0 +1,49 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" A filter to be used with FilteredTraitObserver for observing traits
+with a given metadata. The filter is defined here so that two
+observers created using the same parameters compare equally.
+"""
+
+
+class MetadataFilter:
+    """ Callable to be used with FilteredTraitObserver for filtering traits
+    with the given metadata name.
+
+    This filter returns true if the metadata is defined on the trait, false
+    if it is undefined.
+
+    Attributes
+    ----------
+    metadata_name : str
+        Name of the metadata to filter traits with.
+    """
+
+    def __init__(self, metadata_name):
+        self.metadata_name = metadata_name
+
+    def __call__(self, name, trait):
+        return self.metadata_name in trait.__dict__
```

Comment:
I don't think this is the right test. What's in the `__dict__` is an implementation detail; what matters is what happens on metadata retrieval.

For example, I'd expect the following two traits to behave the same under any metadata filtering:
```
>>> tr1 = Float(2.3, edible=None).as_ctrait()
>>> tr2 = Float(2.3).as_ctrait()
>>> tr2.edible == tr1.edible
True
```

So I think we want to retrieve `getattr(trait, metadata_name)` and then filter on the result. What's not clear to me is whether that filtering should be `is not None` or `bool`. What does the usual `on_trait_change` expression do with this?

Reply:
Ah, looks like [this](https://github.com/enthought/traits/blob/486a28956b077795b7becffc358d5f7a8f48a838/traits/traits_listener.py#L320) is the relevant code. So we want an `is not None` test.

Reply:
Okay. I thought the behaviour of returning `None` for an undefined metadata is conflating the concept of something being undefined VS something being defined with a value of `None` and here we only want to check for metadata existence. So it seems that the fact that these two concepts being combined is in fact the desired behaviour for metadata?

Reply:
Not sure about "desired", but it _is_ the current design. Changing that design would be an orthogonal discussion, I think.

Reply:
I see. Thanks. I will add a comment because this current design is still strange to me, not sure if I am alone in that.